rust/projects/file-io/src/kv.rs

```diff
@@ -63,31 +59,137 @@ impl KvStore {
+            let len = io::copy(&mut entry_reader, &mut new_writer)?;
+            new_index.insert(key.clone(), (new_pos..new_pos + len).into());
+            new_pos += len;
+        }
+        drop(new_writer);
+
+        // As all entries are written to the log, we can safely rename it to a valid log file name
+        let log_path = new_log_path(&self.path);
+        fs::rename(tmp_log_path, &log_path)?;
+
+        // Reopen using the new file name
+        let mut kv_log = KvLog::open(&log_path)?;
+        // Use the index map built on writing instead of reloading the log file
+        kv_log.index = new_index;
+        // Update the KvLog we are using
+        mem::swap(&mut self.kv_log, &mut kv_log);
+
+        // Close old log file before removing it. (It's a must on Windows I think)
+        let old_path = kv_log.path.clone();
+        drop(kv_log);
+        fs::remove_file(old_path)?;
+
+        Ok(())
+    }
+}
+
+struct KvLog {
+    path: PathBuf,
+    reader: BufReaderWithPos<File>,
+    writer: BufWriterWithPos<File>,
+    // stores keys and the pos of the last command to modify each
+    index: HashMap<String, CommandPos>,
+    uncompacted: u64,
+}
+
+impl KvLog {
+    // Pay attention that it does not load the log file automatically
+    fn open(path: impl Into<PathBuf>) -> Result<Self> {
+        let path = path.into();
+        let mut writer = BufWriterWithPos::new(
+            OpenOptions::new()
+                .create(true)
+                .read(true)
+                .append(true)
+                .open(&path)?,
+        )?;
+        // Set pos to end of file
+        writer.seek(SeekFrom::End(0))?;
+
+        let reader = BufReaderWithPos::new(File::open(&path)?)?;
+
+        Ok(KvLog {
+            path,
+            reader,
+            writer,
+            index: HashMap::new(),
+            uncompacted: 0,
+        })
     }
 
-    fn load_from_wal(wal_path: impl AsRef<Path>) -> Result<HashMap<String, String>> {
-        let mut map = HashMap::new();
-        let reader = BufReader::new(File::open(wal_path)?);
-        let stream = Deserializer::from_reader(reader).into_iter::<SetCommand>();
-        for set_cmd in stream {
-            let set_cmd = set_cmd?;
-            map.insert(set_cmd.key, set_cmd.value);
+    fn set(&mut self, key: String, value: String) -> Result<()> {
+        let cmd = SetCommand::new(key, value);
+        let pos = self.writer.pos;
+        serde_json::to_writer(&mut self.writer, &cmd)?;
+        self.writer.flush()?;
+        if let Some(old_cmd) = self.index.insert(cmd.key, (pos..self.writer.pos).into()) {
+            self.uncompacted += old_cmd.len;
         }
-        Ok(map)
+        Ok(())
+    }
+
+    fn get(&mut self, key: String) -> Result<Option<String>> {
+        if let Some(cmd_pos) = self.index.get(&key) {
+            self.reader.seek(SeekFrom::Start(cmd_pos.pos))?;
+            let cmd_reader = (&mut self.reader).take(cmd_pos.len);
```

Comment:
This take method is nice to know about. Good choice. Keep looking out for opportunities to use this kind of rusty api.